lib/ProxySQL_Statistics.cpp

```diff
@@ -1068,3 +1079,103 @@ void ProxySQL_Statistics::MySQL_Query_Cache_sets(SQLite3_result *resultset) {
 		statsdb_disk->execute(buf);
 	}
 }
+
+bool ProxySQL_Statistics::knows_variable_name(std::string variable_name) const
+{
+	return (variable_name_id_map.find(variable_name) != variable_name_id_map.end());
+}
+
+int64_t ProxySQL_Statistics::get_variable_id_for_name(std::string variable_name) {
+	lock_guard<mutex> lock(mu);
+
+	int64_t variable_id = -1; // Negative value indicates not yet found.
+
+	if (knows_variable_name(variable_name)) {
+		variable_id = variable_name_id_map[variable_name];
+	} else {
+		// No matching variable_id found in map. Try loading from the SQLite lookup table on disk 
+		SQLite3_result *result = NULL;
+		int cols;
+		int affected_rows;
+		char *error = NULL;
+		
+		auto select_var_id = [&]() -> int64_t {
+			string var_id_query = "SELECT variable_id FROM history_mysql_status_variables_lookup WHERE variable_name=\"" + variable_name + "\"";
+			statsdb_disk->execute_statement(var_id_query.c_str(), &error , &cols , &affected_rows , &result);
+
+			if (error) {
+				proxy_error("SQLITE ERROR %s", error);
+				free(error);
+				error = NULL;
+				exit(EXIT_SUCCESS); 
+			} 
+			
+			if (result) {
+				if (result->rows_count > 0) {
+					// matching variable_id for variable_name in lookup table
+					SQLite3_row *r = result->rows[0];
+					return strtoll(r->fields[0], NULL, 10);
+				}
+				delete result;
+				result = NULL;
+			}	
+			return -1;
+		};
+
+		variable_id = select_var_id(); // Check for variable name present in the lookup
+
+		if (variable_id < 0) {
+			// No match found, create a new record in the lookup table and then select the newly generated id
+			string insert_var_query = "INSERT INTO history_mysql_status_variables_lookup(variable_name) VALUES(\"" + variable_name + "\")";
+			if (statsdb_disk->execute(insert_var_query.c_str())) {
+				variable_id = select_var_id();
```

Comment:
Since the last query executed was an `INSERT...` there is no need for performing a second query for getting the `id` from the previous query execution, same result can be achieved via `sqlite3_last_insert_rowid()` in the following way:

```
sqlite3* db = statsdb_disk->get_db();
variable_id = sqlite3_last_insert_rowid(db);
```
This way another query and potential disk access can be avoided.

Reply:
That's how I wrote the code at first, when mentioned to Rene, he said this might causes issues from multi-threaded sqlite calls. I figured mutex lock should do, but he was saying that sqlite might still not lock. So that, if another insertion was done, if the timing was wrong, that the wrong id would be returned. I'd gladly change it back that way if this can be confirmed. That way as well, can remove the lambda above.

Reply:
Ha! Nice one... yes, indeed I think there is no guarantee that this is the only thread accessing `statsdb`, and if that is the case as René suggested, this would return the incorrect result. The lambda should stay for now.